pkg/kv/kvprober/kvprober.go

```diff
@@ -306,11 +308,42 @@ func (p *Prober) Metrics() Metrics {
 	return p.metrics
 }
 
+// LogLeaseholderInfo searches through the given tracing recording for log messages
+// indicating updated range information, extracts range IDs and leaseholder node IDs,
+// and logs this information only if the leaseholder for a range has been updated.
+func (p *Prober) LogLeaseholderInfo(ctx context.Context, recording tracingpb.Recording) {
+	// Find the log message containing information about updated ranges
+	informationLog, isValid := recording.FindLogMessage("received updated range")
```

Comment:
Can you paste the textual representation of the trace into this comment thread, so I can review this code with that in mind?

Reply:
`event:kv/kvclient/kvcoord/dist_sender.go:2872 [n3,kvprober,txn=‹17d32b7a›] received updated range info: [‹desc: r63:/Table/6{1-2} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212389000,1, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r62:/Table/6{0-1} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212290000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r64:/Table/6{2-3} [(n1,s1):1, (n2,s2):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.234352000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING›]`

This is the line I used to get the lease information

Reply:
```
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8       0.000ms      0.000ms    === operation:read probe loop _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [/cockroach.roachpb.Internal/Batch: {count: 6, duration 2ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [dist sender send: {count: 3, duration 2ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.000ms      0.000ms    [txn coordinator send: {count: 1, duration 767µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.021ms      0.021ms        === operation:dist sender send _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.021ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 1ms}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.071ms      0.050ms        event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober] querying next range at /Meta2/‹""›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.105ms      0.034ms        event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober] key: /Meta2/‹""›, desc: r1:‹/{Min-System/NodeLiveness}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.125ms      0.021ms        event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober] r1: sending batch ‹1 Scan› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.137ms      0.011ms        event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.144ms      0.007ms        event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.146ms      0.002ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.321ms      0.175ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Scan [/Meta2/"",/Meta2/"\xff\xff\x00"), [max_span_request_keys: 100], [target_bytes: 0]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.332ms      0.011ms            event:server/node.go:1407 [n1] node received request: ‹1 Scan›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.353ms      0.021ms            event:kv/kvserver/store_send.go:154 [n1,s1] executing Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 100], [target_bytes: 0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.360ms      0.007ms            event:kv/kvserver/replica_send.go:176 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.367ms      0.007ms            event:kv/kvserver/concurrency/concurrency_manager.go:214 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.371ms      0.004ms            event:kv/kvserver/concurrency/concurrency_manager.go:295 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.377ms      0.006ms            event:kv/kvserver/concurrency/concurrency_manager.go:339 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.406ms      0.029ms            event:kv/kvserver/replica_read.go:313 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] can drop latches early for batch (Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 100], [target_bytes: 0]); scanning lock table first to detect conflicts
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.463ms      0.057ms            event:kv/kvserver/replica_read.go:125 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] lock table scan complete without conflicts; dropping latches early
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.484ms      0.021ms            event:kv/kvserver/replica_read.go:426 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.580ms      0.096ms            event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] evaluated Scan command ‹header:<kvnemesis_seq:<> key:"\003" end_key:"\003\377\377\000" > ›, txn=<nil> : resp=‹num_keys:68 num_bytes:4031 ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.605ms      0.025ms            event:n1 scan stats: stepped 68 times (130 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 130, key-bytes 1.6 KiB, value-bytes 5.0 KiB, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.616ms      0.011ms            event:kv/kvserver/replica_read.go:238 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.625ms      0.009ms            event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.584ms     -0.041ms            structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","numInterfaceSteps":"68","numInternalSteps":"130","keyBytes":"1614","valueBytes":"5104","pointCount":"130","numScans":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.862ms      0.719ms        event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.888ms      0.026ms        === operation:dist sender send _verbose:‹1› node:‹3› kvprober:
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.888ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 528µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.898ms      0.010ms        event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober] querying next range at /Meta2/‹""›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.914ms      0.016ms        event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober] key: /Meta2/‹""›, desc: r1:‹/{Min-System/NodeLiveness}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.924ms      0.010ms        event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober] r1: sending batch ‹1 Scan› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.929ms      0.005ms        event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.934ms      0.005ms        event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     0.939ms      0.005ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.024ms      0.086ms            === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Scan [/Meta2/"",/Meta2/"\xff\xff\x00"), [max_span_request_keys: 32], [target_bytes: 0]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.030ms      0.006ms            event:server/node.go:1407 [n1] node received request: ‹1 Scan›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.042ms      0.012ms            event:kv/kvserver/store_send.go:154 [n1,s1] executing Scan [/Meta2/‹""›,/Meta2/‹"\xff\xff\x00"›), [max_span_request_keys: 32], [target_bytes: 0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.048ms      0.006ms            event:kv/kvserver/replica_send.go:176 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.055ms      0.007ms            event:kv/kvserver/concurrency/concurrency_manager.go:217 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.071ms      0.016ms            event:kv/kvserver/concurrency/concurrency_manager.go:289 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.083ms      0.012ms            event:kv/kvserver/concurrency/concurrency_manager.go:335 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] optimistically scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.091ms      0.008ms            event:kv/kvserver/replica_read.go:426 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.145ms      0.054ms            event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] evaluated Scan command ‹header:<kvnemesis_seq:<> key:"\003" end_key:"\003\377\377\000" > ›, txn=<nil> : resp=‹resume_span:<key:"\003\247" end_key:"\003\377\377\000" > resume_reason:RESUME_KEY_LIMIT num_keys:32 num_bytes:1935 ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.158ms      0.013ms            event:n1 scan stats: stepped 64 times (175 internal); seeked 2 times (2 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 177, key-bytes 3.9 KiB, value-bytes 7.3 KiB, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 0 gets, 1 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.175ms      0.017ms            event:kv/kvserver/replica_read.go:238 [n1,s1,r1/1:‹/{Min-System/NodeL…}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.183ms      0.008ms            event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.147ms     -0.036ms            structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"2","numInternalSeeks":"2","numInterfaceSteps":"64","numInternalSteps":"175","keyBytes":"4013","valueBytes":"7424","pointCount":"177","numScans":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.326ms      0.393ms        event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.051ms        === operation:txn coordinator send _verbose:‹1› node:‹3› kvprober: txnID:‹17d32b7a-cb42-4e2c-8738-798ec2f7e1fb›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.000ms        [dist sender send: {count: 1, duration 740µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.378ms      0.000ms        [/cockroach.roachpb.Internal/Batch: {count: 2, duration 630µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.382ms      0.004ms            === operation:dist sender send _verbose:‹1› node:‹3› kvprober: txn:‹17d32b7a›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.382ms      0.000ms            [/cockroach.roachpb.Internal/Batch: {count: 2, duration 630µs}]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.394ms      0.012ms            event:kv/kvclient/kvcoord/range_iter.go:188 [n3,kvprober,txn=‹17d32b7a›] querying next range at /Table/61
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.411ms      0.018ms            event:kv/kvclient/kvcoord/range_iter.go:225 [n3,kvprober,txn=‹17d32b7a›] key: /Table/61, desc: r63:‹/Table/6{1-2}› [(n1,s1):1, next=2, gen=0]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.438ms      0.027ms            event:kv/kvclient/kvcoord/dist_sender.go:2579 [n3,kvprober,txn=‹17d32b7a›] routing to nearest replica; leaseholder not known
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.447ms      0.009ms            event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober,txn=‹17d32b7a›] r63: sending batch ‹1 Get› to (n1,s1):1
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.454ms      0.006ms            event:rpc/nodedialer/nodedialer.go:147 [n3,kvprober,txn=‹17d32b7a›] sending request to ‹localhost:26257›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.460ms      0.006ms            event:kv/kvclient/kvcoord/transport.go:213 [n3,kvprober,txn=‹17d32b7a›] sending batch request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.461ms      0.001ms                === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹3› kvprober: txn:‹17d32b7a› span.kind:‹client›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.590ms      0.129ms                === operation:/cockroach.roachpb.Internal/Batch _verbose:‹1› node:‹1› span.kind:‹server› request:‹Get [/Local/Range/Table/61/RangeProbe], [txn: 17d32b7a], [can-forward-ts]›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.600ms      0.010ms                event:server/node.go:1407 [n1] node received request: ‹1 Get›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.620ms      0.020ms                event:kv/kvserver/store_send.go:154 [n1,s1] executing Get [/Local/Range/Table/61/‹RangeProbe›], [txn: 17d32b7a], [can-forward-ts]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.632ms      0.012ms                event:kv/kvserver/replica_send.go:176 [n1,s1,r63/1:‹/Table/6{1-2}›] read-only path
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.640ms      0.008ms                event:kv/kvserver/concurrency/concurrency_manager.go:214 [n1,s1,r63/1:‹/Table/6{1-2}›] sequencing request
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.645ms      0.005ms                event:kv/kvserver/concurrency/concurrency_manager.go:295 [n1,s1,r63/1:‹/Table/6{1-2}›] acquiring latches
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.651ms      0.006ms                event:kv/kvserver/concurrency/concurrency_manager.go:339 [n1,s1,r63/1:‹/Table/6{1-2}›] scanning lock table for conflicting locks
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.666ms      0.015ms                event:kv/kvserver/replica_read.go:313 [n1,s1,r63/1:‹/Table/6{1-2}›] can drop latches early for batch (Get [/Local/Range/Table/61/‹RangeProbe›], [txn: 17d32b7a], [can-forward-ts]); scanning lock table first to detect conflicts
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.677ms      0.011ms                event:kv/kvserver/replica_read.go:125 [n1,s1,r63/1:‹/Table/6{1-2}›] lock table scan complete without conflicts; dropping latches early
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.684ms      0.007ms                event:kv/kvserver/replica_read.go:426 [n1,s1,r63/1:‹/Table/6{1-2}›] executing read-only batch
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.720ms      0.036ms                event:kv/kvserver/replica_evaluate.go:516 [n1,s1,r63/1:‹/Table/6{1-2}›] evaluated Get command ‹header:<kvnemesis_seq:<> key:"\001k\022\305\000\001prbe" > ›, txn="unnamed" meta={id=17d32b7a key=/Min iso=Serializable pri=0.01077494 epo=0 ts=1719946576.032320000,0 min=1719946576.032320000,0 seq=0} lock=false stat=PENDING rts=1719946576.032320000,0 wto=false gul=1719946576.532320000,0 : resp=‹header:<> ›, err=<nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.730ms      0.010ms                event:n1 scan stats: stepped 0 times (0 internal); seeked 1 times (1 internal); block-bytes: (total 0 B, cached 0 B, duration 0s); points: (count 0, key-bytes 0 B, value-bytes 0 B, tombstoned: 0) ranges: (count 0), (contained-points 0, skipped-points 0) evaluated requests: 1 gets, 0 scans, 0 reverse scans
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.739ms      0.009ms                event:kv/kvserver/replica_read.go:238 [n1,s1,r63/1:‹/Table/6{1-2}›] read completed
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.751ms      0.012ms                event:kv/kvserver/replica_send.go:331 [n1,s1,r63/1:‹/Table/6{1-2}›] client had stale range info; returning an update
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.772ms      0.021ms                event:server/node.go:1501 [n1] node sending response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.721ms     -0.051ms                structured:‹{"@type":"type.googleapis.com/cockroach.roachpb.ScanStats","numInterfaceSeeks":"1","numInternalSeeks":"1","numGets":"1","nodeId":1}›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.923ms      0.463ms            event:kv/kvclient/kvcoord/transport.go:215 [n3,kvprober,txn=‹17d32b7a›] received batch response
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.025ms      0.102ms            event:kv/kvclient/kvcoord/dist_sender.go:2872 [n3,kvprober,txn=‹17d32b7a›] received updated range info: [‹desc: r63:/Table/6{1-2} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212389000,1, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r62:/Table/6{0-1} [(n1,s1):1, (n3,s3):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.212290000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING› ‹desc: r64:/Table/6{2-3} [(n1,s1):1, (n2,s2):2, next=3, gen=2], lease: repl=(n1,s1):1 seq=1 start=0,0 epo=1 pro=1719946566.234352000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING›]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.053ms      0.028ms            event:kv/kvclient/rangecache/range_cache.go:1022 [n3,kvprober,txn=‹17d32b7a›] evict cached descriptor: desc=‹desc:r63:/Table/6{1-2} [(n1,s1):1, next=2, gen=0], lease:<empty>›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.062ms      0.009ms            event:kv/kvclient/rangecache/range_cache.go:526 [n3,kvprober,txn=‹17d32b7a›] evicting cached range descriptor with 3 replacements
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.082ms      0.019ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r63:‹/Table/6{1-2}› [(n1,s1):1, (n3,s3):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.099ms      0.018ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r62:‹/Table/6{0-1}› [(n1,s1):1, (n3,s3):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.119ms      0.020ms            event:kv/kvclient/rangecache/range_cache.go:1196 [n3,kvprober,txn=‹17d32b7a›] clearing entries overlapping r64:‹/Table/6{2-3}› [(n1,s1):1, (n2,s2):2, next=3, gen=2]
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.143ms      0.766ms        event:kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:661 [n3,kvprober,txn=‹17d32b7a›] recording span to refresh: ‹/Local/Range/Table/61/RangeProbe›
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.164ms      2.164ms    event:kv/txn.go:1062 [n3,kvprober] kv.Txn did AutoCommit. err: <nil>
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     2.181ms      0.017ms    event:kv/kvprober/kvprober.go:470 [n3,kvprober] kv.Get(/Local/Range/Table/61/‹RangeProbe›), r=63 returned success in 797.417µs

```

This is the entire trace

Reply:
I think you should pull the leaseholder out of this part of the trace:

```
I240702 18:56:16.033155 1350 2@kv/kvprober/kvprober.go:380 ⋮ [T1,Vsystem,n3,kvprober] 8 +     1.447ms      0.009ms            event:kv/kvclient/kvcoord/dist_sender.go:2644 [n3,kvprober,txn=‹17d32b7a›] r63: sending batch ‹1 Get› to (n1,s1):1
```

We want the final occurrence of above in the trace. The earlier occurrences are related to the planning happening in `planner.go`, instead of the probing of the target range happening in `kvprober.go`'s `readProbeImpl`, etc. methods. We are interested in the leaseholder of the target range specifically.